src/Elasticsearch.Net/Responses/Dynamic/DynamicValue.cs

```diff
@@ -45,19 +47,32 @@ public bool HasValue
 			}
 		}
 
+		public static DynamicValue NullValue { get; } = new DynamicValue(null);
+		public static DynamicValue SelfOrNew(object v) => v is DynamicValue av ? av : new DynamicValue(v);
+
 		public DynamicValue this[int i]
 		{
 			get
 			{
-				if (!HasValue)
-					return new DynamicValue(null);
+				if (!HasValue) return NullValue;
+
+				var v = Value;
+
+				if (v is IList l && l.Count - 1 >= i)
+					return SelfOrNew(l[i]);
+				if (v is IList<object> o && o.Count - 1 >= i)
+					return SelfOrNew(o[i]);
 
-				var l = Value as IList;
-				if (l != null && l.Count - 1 >= i)
+				if (v is IDictionary<string, object> d)
 				{
-					return new DynamicValue(l[i]);
+					if (d.TryGetValue(i.ToString(CultureInfo.InvariantCulture), out v))
+						return SelfOrNew(v);
+
+					if (i >= d.Count) return new DynamicValue(null);
+					var at = d[d.Keys.ElementAt(i)];
+					return SelfOrNew(at);
 				}
-				return new DynamicValue(null);
+				return NullValue;
```

Comment:
Is is safe to return the same instance of `new DynamicValue(null)` each time?

Reply:
yes, its not possible to mutate the `null` it holds